usl_models/usl_models/flood_ml/dataset.py

```diff
@@ -0,0 +1,231 @@
+    for t in range(T_max):
+        window_input = model.Input(
+            geospatial=model_input["geospatial"],
+            temporal=_extract_temporal(t, n_flood_maps, model_input["temporal"]),
+            spatiotemporal=_extract_spatiotemporal(t, n_flood_maps, labels),
+        )
+        yield window_input, labels[t]
+
+
+def _extract_temporal(t: int, n: int, temporal: tf.Tensor) -> tf.Tensor:
+    (_, D) = temporal.shape
+    zeros = tf.zeros(shape=(max(n - t, 0), D))
+    data = temporal[max(t - n, 0) : t]
+    return tf.concat([zeros, data], axis=0)
+
+
+def _extract_spatiotemporal(t: int, n: int, labels: tf.Tensor) -> tf.Tensor:
+    """Extracts spatiotemporal tensor from labeled data.
+
+    Args:
+      t: The current timestep.
+      n: The window size.
+      labels: The training labels.
+
+    Returns:
+      The slice labels[t-n: t] with zero padding so the output tensor is always
+      of shape (n, H, W, 1).
+    """
+    (_, H, W, *_) = labels.shape
+    zeros = tf.zeros(shape=(max(n - t, 0), H, W), dtype=tf.float32)
+    data = labels[max(t - n, 0) : t]
+    return tf.concat([zeros, data], axis=0)
+
+
+def _iter_model_inputs(
+    firestore_client: firestore.Client,
+    storage_client: storage.Client,
+    sim_name: str,
+    n_flood_maps: int,
+    m_rainfall: int,
+    max_chunks: Optional[int],
+) -> Iterator[Tuple[model.Input, tf.Tensor]]:
+    """Yields model inputs for each spatial chunk in the simulation."""
+    temporal = _generate_temporal_tensor(
+        firestore_client, storage_client, sim_name, m_rainfall
+    )
+    feature_label_gen = _iter_feature_label_tensors(
+        firestore_client, storage_client, sim_name
+    )
+
+    for i, (geospatial, labels) in enumerate(feature_label_gen):
+        if max_chunks is not None and i >= max_chunks:
+            return
+
+        model_input = model.Input(
+            temporal=temporal,
+            geospatial=geospatial,
+            spatiotemporal=tf.zeros(
+                shape=(
+                    n_flood_maps,
+                    constants.MAP_HEIGHT,
+                    constants.MAP_WIDTH,
+                    1,
+                )
+            ),
+        )
+        yield model_input, labels
+
+
+def _generate_temporal_tensor(
+    firestore_client: firestore.Client,
+    storage_client: storage.Client,
+    sim_name: str,
+    m_rainfall: int,
+) -> tf.Tensor:
+    """Creates a temporal tensor from the numpy array stored in GCS."""
+    temporal_metadata = metastore.get_temporal_feature_metadata(
+        firestore_client, sim_name
+    )
+    gcs_url = temporal_metadata["as_vector_gcs_uri"]
+
+    logging.info("Retrieving temporal features from %s.", gcs_url)
+    temporal_array = _download_as_array(storage_client, gcs_url)
+    temporal_tensor = tf.convert_to_tensor(
+        numpy.tile(temporal_array, (m_rainfall, 1)).T,
+        dtype=tf.float32,
+    )
+    return temporal_tensor
+
+
+def _iter_feature_label_tensors(
```

Comment:
Perhaps it's more clear to replace "feature" with "geo_feature" in this function and its callsite.